packages/x-data-grid-pro/src/hooks/features/dataSource/useGridDataSource.ts

```diff
@@ -0,0 +1,271 @@
+        return;
+      }
+
+      nestedDataManager.clear();
+      scheduledGroups.current = 0;
+      const dataSourceState = privateApiRef.current.state.dataSource;
+      if (dataSourceState !== INITIAL_STATE) {
+        privateApiRef.current.resetDataSourceState();
+      }
+
+      const fetchParams = gridGetRowsParamsSelector(privateApiRef);
+
+      const cachedData = privateApiRef.current.unstable_dataSourceCache?.get(fetchParams);
+
+      if (cachedData != null) {
+        const rows = cachedData.rows;
+        privateApiRef.current.setRows(rows);
+        if (cachedData.rowCount) {
+          privateApiRef.current.setRowCount(cachedData.rowCount);
+        }
+        return;
+      }
+
+      const isLoading = gridRowsLoadingSelector(privateApiRef);
+      if (!isLoading) {
+        privateApiRef.current.setLoading(true);
+      }
+
+      try {
+        const getRowsResponse = await getRows(fetchParams);
+        privateApiRef.current.unstable_dataSourceCache?.set(fetchParams, getRowsResponse);
+        if (getRowsResponse.rowCount) {
+          privateApiRef.current.setRowCount(getRowsResponse.rowCount);
+        }
+        privateApiRef.current.setRows(getRowsResponse.rows);
+        privateApiRef.current.setLoading(false);
+      } catch (error) {
+        privateApiRef.current.setRows([]);
+        privateApiRef.current.setLoading(false);
+        onError?.(error as Error, fetchParams);
+      }
+    },
+    [nestedDataManager, privateApiRef, props.unstable_dataSource?.getRows, onError],
+  );
+
+  const fetchRowChildren = React.useCallback<GridDataSourcePrivateApi['fetchRowChildren']>(
+    async (id) => {
+      if (!props.treeData) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+      const getRows = props.unstable_dataSource?.getRows;
+      if (!getRows) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+
+      const rowNode = privateApiRef.current.getRowNode(id) as GridServerSideGroupNode;
+      if (!rowNode) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+
+      const fetchParams = { ...gridGetRowsParamsSelector(privateApiRef), groupKeys: rowNode.path };
+
+      const cachedData = privateApiRef.current.unstable_dataSourceCache?.get(fetchParams);
+
+      if (cachedData != null) {
+        const rows = cachedData.rows;
+        nestedDataManager.setRequestSettled(id);
+        privateApiRef.current.updateServerRows(rows, rowNode.path);
+        if (cachedData.rowCount) {
+          privateApiRef.current.setRowCount(cachedData.rowCount);
+        }
+        privateApiRef.current.setRowChildrenExpansion(id, true);
+        privateApiRef.current.unstable_dataSource.setChildrenLoading(id, false);
+        return;
+      }
+
+      const existingError = gridDataSourceErrorsSelector(privateApiRef)[id] ?? null;
+      if (existingError) {
+        privateApiRef.current.unstable_dataSource.setChildrenFetchError(id, null);
```

Comment:
Technically speaking, clearing the state by setting it to `null` is going to leak memory as those keys are going to stay in the state forever.